File: Backend/src/services/unsplash.service.ts
import axios from 'axios';
import chalk from 'chalk';
import dotenv from 'dotenv';
import { config } from '../config/config';

dotenv.config();

/**
 * Fetches a high-quality existing image URL from Unsplash for a given place name.
 * Uses the Unsplash Search API.
 * @param query The name of the place to search for (e.g., "Paris", "Kyoto").
 * @returns A string URL of the image, or undefined if not found.
 */
export const fetchUnsplashImage = async (query: string): Promise<string | undefined> => {
    try {
        const accessKey = config.UNSPLASH_ACCESS_KEY || process.env.UNSPLASH_ACCESS_KEY;

        if (!accessKey) {
            console.error(chalk.red("Unsplash API Key is missing! Please set UNSPLASH_ACCESS_KEY in .env"));
            return undefined;
        }

        const endpoint = `https://api.unsplash.com/search/photos`;

        const params = {
            query: query,
            per_page: 1,
            orientation: 'landscape',
            client_id: accessKey
        };

        const response = await axios.get(endpoint, { params });

        const results = response.data?.results;

        if (!results || results.length === 0) {
            console.log(chalk.yellow(`No image found on Unsplash for ${query}`));
            return undefined;
        }

        const imageUrl = results[0].urls?.regular;

        if (imageUrl) {
            console.log(chalk.green(`Image fetched from Unsplash for ${query}`));
            return imageUrl;
        } else {
            return undefined;
        }

    } catch (error) {
        console.error(chalk.red(`Error fetching Unsplash image for ${query}:`, error instanceof Error ? error.message : error));
        return undefined;
    }
};

File: Backend/src/services/unsplash_service.py
import httpx
from typing import Optional, List
from src.core.config import settings
import logging

logger = logging.getLogger(__name__)

async def fetch_unsplash_image(query: str) -> Optional[str]:
    """
    Fetches a high-quality existing image URL from Unsplash for a given place name.
    """
    access_key = settings.UNSPLASH_ACCESS_KEY
    if not access_key:
        logger.error("Unsplash API Key is missing!")
        return None

    endpoint = "https://api.unsplash.com/search/photos"
    params = {
        "query": query,
        "per_page": 1,
        "orientation": "landscape",
        "client_id": access_key
    }

    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(endpoint, params=params)
            response.raise_for_status()
            data = response.json()
            
            results = data.get("results", [])
            if not results:
                return None
                
            return results[0].get("urls", {}).get("regular")
    except Exception as e:
        logger.error(f"Error fetching Unsplash image for {query}: {e}")
        return None

async def fetch_unsplash_images(query: str, count: int = 10) -> List[str]:
    """
    Fetches multiple images from Unsplash.
    """
    access_key = settings.UNSPLASH_ACCESS_KEY
    if not access_key:
        logger.error("Unsplash API Key is missing!")
        return []

    endpoint = "https://api.unsplash.com/search/photos"
    params = {
        "query": query,
        "per_page": count,
        "orientation": "landscape",
        "client_id": access_key
    }

    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(endpoint, params=params)
            response.raise_for_status()
            data = response.json()
            
            results = data.get("results", [])
            if not results:
                return []
                
            return [img.get("urls", {}).get("small") for img in results if img.get("urls", {}).get("small")]
    except Exception as e:
        logger.error(f"Error fetching Unsplash images for {query}: {e}")
        return []
